File: Code/Chapter 13/C#/SqlServerTVF/NameToAscii.cs
using System;
using System.Data;
using System.Data.Sql;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;
using System.Collections;

namespace SqlServerTVF
{
   public partial class UserDefinedFunctions
   {
      [Microsoft.SqlServer.Server.SqlFunction(FillRowMethodName = "FillRow",
      TableDefinition = "charpart nchar(1), intpart int")]
      public static IEnumerable NameToAscii(string InputName)
      {
         return new NameSplitter(InputName.ToCharArray());
      }

      //[Microsoft.SqlServer.Server.SqlFunction(FillRowMethodName = "FillRow", 
      //   TableDefinition="charpart nchar(1), intpart int")]
      //public static IEnumerable NameToAscii(string InputName)
      //{
      //   foreach (char c in InputName)
      //   {
      //      yield return new NameRow(c, (int)c);
      //   }
      //}

      public static void FillRow(object row, out char charpart, out int intpart)
      {
         // break the row into its columnar parts. 
         charpart = ((NameRow)row).CharPart;
         intpart = ((NameRow)row).IntPart;
      }
   };

   public class NameRow
   {
      public Char CharPart;
      public Int32 IntPart;

      public NameRow(Char c, Int32 i)
      {
         CharPart = c;
         IntPart = i;
      }
   }

   public class NameSplitter :  IEnumerable, IEnumerator
   {
	   private Int32 idx = - 1;
	   private Char[] _inputName;

	   public NameSplitter(Char[] InputName)
	   {
		   _inputName = InputName;
	   }

	   public System.Collections.IEnumerator GetEnumerator()
	   {
		   return new NameSplitter(_inputName);
	   }

	   public object Current
	   {
		   get
		   {
			   if(( idx > - 1 ))
			   {
				   Char c = _inputName[idx];
				   return new NameRow(c, Convert.ToInt32(c));
			   }
			   else
			   {
				   return - 1;
			   }
		   }
	   }
	   public bool MoveNext()
	   {
		   idx = idx + 1;
		   if(( idx < _inputName.Length ))
		   {
			   return true;
		   }
		   else
		   {
			   return false;
		   }
	   }

	   public void Reset()
	   {
		   idx = - 1;
	   }
   }
}

//Create Assembly SqlServerTVF
//from
//'C:\PlayArea\SqlServerTVF\bin\Debug\SqlServerTVF.dll'
//GO

//Create Function NameToAscii 
//(
//   @InputName NVARCHAR(4000)
//)
//Returns Table 
//(
//   charpart nchar, 
//   intpart Int
//)
//As
//External Name
//SqlServerTVF.[SqlServerTVF.UserDefinedFunctions].NameToAscii 
//Go

//Select * from dbo.NameToAscii ('Sahil Malik') order by intpart
//Go

File: Code/Chapter 13/VB.NET/SqlServerTVF/NameToAscii.vb
Imports System
Imports System.Data
Imports System.Data.Sql
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server
Imports System.Collections
Imports System.Runtime.InteropServices

Namespace SqlServerTVF
   Partial Public Class UserDefinedFunctions
      <Microsoft.SqlServer.Server.SqlFunction(FillRowMethodName:="FillRow", _
         TableDefinition:="charpart nchar(1), intpart int")> _
      Public Shared Function NameToAscii(ByVal InputName As String) As IEnumerable
         Return New NameSplitter(InputName.ToCharArray())
      End Function

      Public Shared Sub FillRow(ByVal row As Object, _
         <Out()> ByRef charpart As Char, <Out()> ByRef intpart As Integer)
         ' break the row into its columnar parts. 
         charpart = CType(row, NameRow).CharPart
         intpart = CType(row, NameRow).IntPart
      End Sub
   End Class


   Public Class NameSplitter
      Implements IEnumerable, IEnumerator

      Private idx As Int32 = -1
      Private _inputName As Char()
      Private _current As Char

      Public Sub New(ByVal InputName As Char())
         _inputName = InputName
      End Sub

      Public Function GetEnumerator() As System.Collections.IEnumerator _
         Implements System.Collections.IEnumerable.GetEnumerator
         Return New NameSplitter(_inputName)
      End Function

      Public ReadOnly Property Current() As Object _
         Implements System.Collections.IEnumerator.Current
         Get
            If (idx > -1) Then
               Dim c As Char = _inputName(idx)
               Return New NameRow(c, Asc(c))
            Else
               Return -1
            End If
         End Get
      End Property

      Public Function MoveNext() As Boolean _
         Implements System.Collections.IEnumerator.MoveNext
         idx = idx + 1
         If (idx < _inputName.Length) Then
            Return True
         Else
            Return False
         End If
      End Function

      Public Sub Reset() Implements System.Collections.IEnumerator.Reset
         idx = -1
      End Sub
   End Class

   Public Class NameRow
      Public CharPart As Char
      Public IntPart As Int32

      Public Sub New(ByVal c As Char, ByVal i As Int32)
         CharPart = c
         IntPart = i
      End Sub
   End Class
End Namespace